<a href="https://colab.research.google.com/github/kondyskrzysiek/API/blob/master/141_150_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 141](#1)
* [Ćwiczenie 142](#2)
* [Ćwiczenie 143](#3)
* [Ćwiczenie 144](#4)
* [Ćwiczenie 145](#5)
* [Ćwiczenie 146](#6)
* [Ćwiczenie 147](#7)
* [Ćwiczenie 148](#8)
* [Ćwiczenie 149](#9)
* [Ćwiczenie 150](#10)

### <a name='0'></a>Import biblioteki

In [1]:
import numpy as np
import pandas as pd

np.random.seed(42)
np.__version__

'1.21.5'

### <a name='1'></a> Ćwiczenie 141
Podane są dwa obiekty typu _Series_:

```
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
```
Połącz te dwa obiekty w jeden obiekt typu _DataFrame_ (jako dwie kolumny) i przypisz do zmiennej _df_. Nadaj nazwy kolumn odpowiednio _col1_ oraz _col2_.


In [37]:
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))

df = pd.DataFrame(data=pd.concat([s1,s2],axis=1))
df.columns = ['col1','col2']
df

col1      col2
0   0.652746 -0.522723
1   0.238183  1.049009
2   0.099441 -0.704344
3   0.243172 -1.408461
4   0.722267 -1.556629
5   0.855696  0.606010
6   0.830220 -1.280429
7   0.397184  1.754794
8   0.668085 -2.081929
9   0.204984  1.696456
10  0.293148  0.211017
11  0.896336 -0.096713
12  0.013002 -0.544919
13  0.085509  0.399136
14  0.207886 -0.037635
15  0.026532  1.103302
16  0.181435  0.114228
17  0.583042  0.150302
18  0.421425 -0.363612
19  0.892672 -0.056946

### <a name='2'></a> Ćwiczenie 142
Wytnij z obiektu _df_ wiersze, dla których wartość w kolumnie _col2_ jest pomiędzy 0.0 a 1.0 (włącznie).

In [38]:
df[df['col2'].between(0.0,1.0)]

col1      col2
5   0.855696  0.606010
10  0.293148  0.211017
13  0.085509  0.399136
16  0.181435  0.114228
17  0.583042  0.150302

In [39]:
df[(df['col2'] >= 0.0) & (df['col2'] <= 1.0)]

col1      col2
5   0.855696  0.606010
10  0.293148  0.211017
13  0.085509  0.399136
16  0.181435  0.114228
17  0.583042  0.150302

### <a name='3'></a> Ćwiczenie 143
Przypisz nową kolumnę o nazwie _col3_, która przyjmie wartość 1 w momencie gdy w kolumnie _col2_ jest wartość nieujemna i przeciwnie -1.

In [40]:
df['col3'] = (df['col2'] >= 0.0).map({True:1,False:-1})
df

col1      col2  col3
0   0.652746 -0.522723    -1
1   0.238183  1.049009     1
2   0.099441 -0.704344    -1
3   0.243172 -1.408461    -1
4   0.722267 -1.556629    -1
5   0.855696  0.606010     1
6   0.830220 -1.280429    -1
7   0.397184  1.754794     1
8   0.668085 -2.081929    -1
9   0.204984  1.696456     1
10  0.293148  0.211017     1
11  0.896336 -0.096713    -1
12  0.013002 -0.544919    -1
13  0.085509  0.399136     1
14  0.207886 -0.037635    -1
15  0.026532  1.103302     1
16  0.181435  0.114228     1
17  0.583042  0.150302     1
18  0.421425 -0.363612    -1
19  0.892672 -0.056946    -1

### <a name='4'></a> Ćwiczenie 144
Przypisz nową kolumnę o nazwie _col4_, która przytnie wartości z kolumny _col2_ do przedziału $[-1.0, 1.0]$. 
Innymi słowy, dla wartości poniżej -1.0 należy wstawić wartość -1.0, dla wartości powyżej 1.0 wartość 1.0.


In [42]:
df['col4'] = df['col2'].clip(lower=-1.0,upper=1.0)
df

col1      col2  col3      col4
0   0.652746 -0.522723    -1 -0.522723
1   0.238183  1.049009     1  1.000000
2   0.099441 -0.704344    -1 -0.704344
3   0.243172 -1.408461    -1 -1.000000
4   0.722267 -1.556629    -1 -1.000000
5   0.855696  0.606010     1  0.606010
6   0.830220 -1.280429    -1 -1.000000
7   0.397184  1.754794     1  1.000000
8   0.668085 -2.081929    -1 -1.000000
9   0.204984  1.696456     1  1.000000
10  0.293148  0.211017     1  0.211017
11  0.896336 -0.096713    -1 -0.096713
12  0.013002 -0.544919    -1 -0.544919
13  0.085509  0.399136     1  0.399136
14  0.207886 -0.037635    -1 -0.037635
15  0.026532  1.103302     1  1.000000
16  0.181435  0.114228     1  0.114228
17  0.583042  0.150302     1  0.150302
18  0.421425 -0.363612    -1 -0.363612
19  0.892672 -0.056946    -1 -0.056946

### <a name='5'></a> Ćwiczenie 145
Znajdź 5 największych wartości dla kolumny _col2_.

In [51]:
# np.sort(df['col2'])[::-1][:5]
df['col2'].nlargest(5)

7     1.754794
9     1.696456
15    1.103302
1     1.049009
5     0.606010
Name: col2, dtype: float64

Znajdź 5 najmniejszych wartości dla kolumny _col2_.

In [57]:
# np.sort(df['col2'])[:5]
df['col2'].nsmallest(5)

8   -2.081929
4   -1.556629
3   -1.408461
6   -1.280429
2   -0.704344
Name: col2, dtype: float64

### <a name='6'></a> Ćwiczenie 146
Wyznacz skumulowaną sumę dla każdej kolumny.

__Wskazówka:__ Użyj metody _pd.DataFrame.cumsum()_.

In [61]:
df.cumsum()

col1      col2  col3      col4
0   0.652746 -0.522723    -1 -0.522723
1   0.890929  0.526286     0  0.477277
2   0.990370 -0.178057    -1 -0.227067
3   1.233542 -1.586519    -2 -1.227067
4   1.955809 -3.143148    -3 -2.227067
5   2.811506 -2.537138    -2 -1.621057
6   3.641726 -3.817567    -3 -2.621057
7   4.038909 -2.062773    -2 -1.621057
8   4.706994 -4.144703    -3 -2.621057
9   4.911979 -2.448246    -2 -1.621057
10  5.205126 -2.237229    -1 -1.410039
11  6.101462 -2.333942    -2 -1.506752
12  6.114464 -2.878861    -3 -2.051671
13  6.199973 -2.479725    -2 -1.652535
14  6.407859 -2.517360    -3 -1.690170
15  6.434391 -1.414058    -2 -0.690170
16  6.615827 -1.299830    -1 -0.575942
17  7.198868 -1.149528     0 -0.425641
18  7.620293 -1.513140    -1 -0.789253
19  8.512964 -1.570086    -2 -0.846199

### <a name='7'></a> Ćwiczenie 147
Wyznacz medianę dla zmiennej _col2_ (inaczej kwantyl rzędu 0.5).

In [62]:
df['col2'].median()

-0.047290163072957914

### <a name='8'></a> Ćwiczenie 148
Wytnij wiersze, dla których zmienna _col2_ przyjmuje wartości większe od 0.0.

In [63]:
df[df['col2'] > 0]

col1      col2  col3      col4
1   0.238183  1.049009     1  1.000000
5   0.855696  0.606010     1  0.606010
7   0.397184  1.754794     1  1.000000
9   0.204984  1.696456     1  1.000000
10  0.293148  0.211017     1  0.211017
13  0.085509  0.399136     1  0.399136
15  0.026532  1.103302     1  1.000000
16  0.181435  0.114228     1  0.114228
17  0.583042  0.150302     1  0.150302

### <a name='9'></a> Ćwiczenie 149
Wytnij 5 pierwszych wierszy obiektu _df_ i przekształć je do słownika.

In [64]:
dict_df = df[:5].to_dict()
dict_df

{'col1': {0: 0.6527461078518747,
  1: 0.23818278104672652,
  2: 0.09944139275934516,
  3: 0.24317219099945409,
  4: 0.7222669318556592},
 'col2': {0: -0.5227230205190395,
  1: 1.0490092258368893,
  2: -0.7043436905427592,
  3: -1.4084612963635597,
  4: -1.5566291735239037},
 'col3': {0: -1, 1: 1, 2: -1, 3: -1, 4: -1},
 'col4': {0: -0.5227230205190395,
  1: 1.0,
  2: -0.7043436905427592,
  3: -1.0,
  4: -1.0}}

### <a name='10'></a> Ćwiczenie 150
Wytnij 5 pierwszych wierszy obiektu _df_, przekształć je do formatu Markdown i przypisz do zmiennej _df_markdown_.

In [68]:
df_markdown = df[:5].to_markdown()

Wydrukuj do konsoli zawartość zmiennej _df_markdown_.

In [69]:
print(df_markdown)

|    |      col1 |      col2 |   col3 |      col4 |
|---:|----------:|----------:|-------:|----------:|
|  0 | 0.652746  | -0.522723 |     -1 | -0.522723 |
|  1 | 0.238183  |  1.04901  |      1 |  1        |
|  2 | 0.0994414 | -0.704344 |     -1 | -0.704344 |
|  3 | 0.243172  | -1.40846  |     -1 | -1        |
|  4 | 0.722267  | -1.55663  |     -1 | -1        |


Skopiuj wynik uruchomienia powyższej komórki i wklej poniżej:

|    |      col1 |      col2 |   col3 |      col4 |
|---:|----------:|----------:|-------:|----------:|
|  0 | 0.652746  | -0.522723 |     -1 | -0.522723 |
|  1 | 0.238183  |  1.04901  |      1 |  1        |
|  2 | 0.0994414 | -0.704344 |     -1 | -0.704344 |
|  3 | 0.243172  | -1.40846  |     -1 | -1        |
|  4 | 0.722267  | -1.55663  |     -1 | -1        |
